src/v/cloud_topics/objects/level_one_object.cc

```diff
@@ -0,0 +1,447 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_topics/objects/level_one_object.h"
+
+#include "base/vassert.h"
+#include "bytes/iostream.h"
+#include "model/fundamental.h"
+#include "model/record.h"
+#include "reflection/type_traits.h"
+#include "serde/rw/rw.h"
+#include "serde/rw/vector.h"
+#include "storage/record_batch_builder.h"
+
+#include <seastar/core/file.hh>
+#include <seastar/core/fstream.hh>
+#include <seastar/core/reactor.hh>
+
+#include <fmt/format.h>
+
+#include <algorithm>
+#include <bit>
+#include <cstring>
+#include <type_traits>
+#include <utility>
+
+namespace experimental::cloud_topics::l1 {
+
+namespace {
+
+// the delimiter for what kind of data is next in the object.
+enum class data_type : uint8_t {
+    kafka_batch = 0,
+    partition_marker = 1,
+    footer = 2,
+};
+
+template<typename T>
+requires std::is_integral_v<T> || std::is_scoped_enum_v<T>
+         || reflection::is_rp_named_type<T>
+         || std::is_same_v<model::timestamp, T>
+         || std::is_same_v<model::record_batch_attributes, T>
+constexpr auto as_bytes(T value) {
+    if constexpr (std::is_scoped_enum_v<T>) {
+        return std::bit_cast<
+          std::array<char, sizeof(std::underlying_type_t<T>)>,
+          std::underlying_type_t<T>>(std::to_underlying(value));
+    } else if constexpr (
+      reflection::is_rp_named_type<T> || std::is_same_v<model::timestamp, T>) {
+        return std::bit_cast<
+          std::array<char, sizeof(typename T::type)>,
+          typename T::type>(value());
+    } else if constexpr (std::is_same_v<model::record_batch_attributes, T>) {
+        return std::bit_cast<
+          std::array<char, sizeof(typename T::type)>,
+          typename T::type>(value.value());
+    } else {
+        return std::bit_cast<std::array<char, sizeof(T)>, T>(value);
+    }
+}
+
+template<typename T>
+requires std::is_integral_v<T> || std::is_scoped_enum_v<T>
+         || reflection::is_rp_named_type<T>
+         || std::is_same_v<model::timestamp, T>
+         || std::is_same_v<model::record_batch_attributes, T>
+constexpr T from_bytes(const void* ptr) {
+    if constexpr (std::is_scoped_enum_v<T>) {
+        std::underlying_type_t<T> value;
+        std::memcpy(&value, ptr, sizeof(value));
+        return static_cast<T>(value);
+    } else if constexpr (
+      reflection::is_rp_named_type<T> || std::is_same_v<model::timestamp, T>
+      || std::is_same_v<model::record_batch_attributes, T>) {
+        typename T::type value;
+        std::memcpy(&value, ptr, sizeof(value));
+        return T(value);
+    } else {
+        T value;
+        std::memcpy(&value, ptr, sizeof(value));
+        return value;
+    }
```

Comment:
don't we need to consider byte ordering? it seems like we'd be better off using serde rather than rolling some new serialization here. easy to change later, for sure so not a blocker, but also surprising to see. is there some motivation for this?

Reply:
yeah I guess serde is fine here and works as it's fixed width for this stuff, parsing does require an iobuf parser and extra machinery. We will have to use read nested and sort of abuse the framework a bit for the read side

Reply:
I added byte ordering: https://github.com/redpanda-data/redpanda/pull/26563/commits/77e7605c0b37ebdcb3cc18e3ad85a57f83cd0bd9

I tried to use serde and it's just pretty darn clunky the forced usage of `iobuf` in places like reading a single byte for the data_type or reading 4 bytes for the size prefix in our partition markers. I added a note to this effect. If you feel strongly we should use serde I'm happy to just do it.